test/e2e-go/upgrades/application_support_test.go

```diff
@@ -0,0 +1,550 @@
+	client.WaitForRound(round + 2)
+
+	// check that the primary node still has this transaction in it's transaction pool.
+	pendingTx, err := client.GetPendingTransactions(1)
+	require.NoError(t, err)
+
+	round, err = client.CurrentRound()
+	require.NoError(t, err)
+	if round > round+primaryNodeUnupgradedProtocol.DefaultUpgradeWaitRounds {
+		t.Skip("Test platform is too slow for this test")
+	}
+
+	require.Equal(t, uint64(1), pendingTx.TotalTxns)
+
+	// check that the secondary node doesn't have that transaction in it's transaction pool.
+	pendingTx, err = secondary.GetPendingTransactions(1)
+	require.NoError(t, err)
+	require.Equal(t, uint64(0), pendingTx.TotalTxns)
+
+	curStatus, err := client.Status()
+	require.NoError(t, err)
+	initialStatus := curStatus
+
+	startLoopTime := time.Now()
+
+	// wait until the network upgrade : this can take a while.
+	for curStatus.LastVersion == initialStatus.LastVersion {
+		curStatus, err = client.Status()
+		require.NoError(t, err)
+
+		require.Less(t, int64(time.Now().Sub(startLoopTime)), int64(3*time.Minute))
+		time.Sleep(time.Duration(roundTimeMs) * time.Millisecond)
+		round = curStatus.LastRound
+	}
+
+	// now, that we have upgraded to the new protocol which supports applications, try again.
+	tx, err = client.FillUnsignedTxTemplate(creator, round, round+100, fee, tx)
+	require.NoError(t, err)
+	signedTxn, err = client.SignTransactionWithWallet(wh, nil, tx)
+	require.NoError(t, err)
+	_, err = client.BroadcastTransaction(signedTxn)
+	require.NoError(t, err)
+
+	curStatus, err = client.Status()
+	require.NoError(t, err)
+
+	round = curStatus.LastRound
+
+	client.WaitForRound(round + 2)
+	pendingTx, err = client.GetPendingTransactions(1)
+	require.NoError(t, err)
+	require.Equal(t, uint64(0), pendingTx.TotalTxns)
+
+	// check creator's balance record for the app entry and the state changes
+	ad, err = client.AccountData(creator)
+	require.NoError(t, err)
+	require.Equal(t, 1, len(ad.AppParams))
+	var appIdx basics.AppIndex
+	var params basics.AppParams
+	for i, p := range ad.AppParams {
+		appIdx = i
+		params = p
+		break
+	}
+	require.Equal(t, approval, params.ApprovalProgram)
+	require.Equal(t, clearstate, params.ClearStateProgram)
+	require.Equal(t, schema, params.LocalStateSchema)
+	require.Equal(t, schema, params.GlobalStateSchema)
+	require.Equal(t, 1, len(params.GlobalState))
+	value, ok := params.GlobalState["counter"]
+	require.True(t, ok)
+	require.Equal(t, uint64(1), value.Uint)
+
+	require.Equal(t, 1, len(ad.AppLocalStates))
+	state, ok := ad.AppLocalStates[appIdx]
+	require.True(t, ok)
+	require.Equal(t, schema, state.Schema)
+	require.Equal(t, 1, len(state.KeyValue))
+	value, ok = state.KeyValue["counter"]
+	require.True(t, ok)
+	require.Equal(t, uint64(1), value.Uint)
+
+	// call the app
+	tx, err = client.MakeUnsignedAppOptInTx(uint64(appIdx), nil, nil, nil)
+	require.NoError(t, err)
+	tx, err = client.FillUnsignedTxTemplate(user, 0, 0, fee, tx)
```

Comment:
why the round is set to 0?

Reply:
Setting it to zero allow the underlying implementation to fill it up for us.